src/grpc/infer_handler.cc

```diff
@@ -0,0 +1,970 @@
+                .c_str());
+      }
+
+      *buffer = const_cast<void*>(pr->second.base_);
+      *actual_memory_type = pr->second.memory_type_;
+      *actual_memory_type_id = pr->second.memory_type_id_;
+
+      LOG_VERBOSE(1) << "GRPC: using shared-memory for '" << tensor_name
+                     << "', size: " << byte_size << ", addr: " << *buffer;
+      return nullptr;  // Success
+    }
+
+    // Not using shared memory so allocate a buffer. The buffer we
+    // create is directly in the response protobuf so we can't
+    // allocate any type other than CPU.
+    //
+    // FIXME we could use pinned CPU memory here.
+    if (*actual_memory_type != TRITONSERVER_MEMORY_CPU) {
+      LOG_VERBOSE(1) << "GRPC: unable to provide '" << tensor_name << "' in "
+                     << TRITONSERVER_MemoryTypeString(*actual_memory_type)
+                     << ", will use "
+                     << TRITONSERVER_MemoryTypeString(TRITONSERVER_MEMORY_CPU);
+      *actual_memory_type = TRITONSERVER_MEMORY_CPU;
+      *actual_memory_type_id = 0;
+    }
+
+    raw_output->resize(byte_size);
+    *buffer = static_cast<void*>(&((*raw_output)[0]));
+
+    LOG_VERBOSE(1) << "GRPC: using buffer for '" << tensor_name
+                   << "', size: " << byte_size << ", addr: " << *buffer;
+  }
+
+  return nullptr;  // Success
+}
+
+void
+ModelInferHandler::Execute(InferHandler::State* state)
+{
+  TRITONSERVER_Error* err = nullptr;
+  const inference::ModelInferRequest& request = state->request_;
+  auto response_queue = state->response_queue_;
+  int64_t requested_model_version;
+  if (err == nullptr) {
+    err = GetModelVersionFromString(
+        request.model_version(), &requested_model_version);
+  }
+
+  if (err == nullptr) {
+    uint32_t txn_flags;
+    err = TRITONSERVER_ServerModelTransactionProperties(
+        tritonserver_.get(), request.model_name().c_str(),
+        requested_model_version, &txn_flags, nullptr /* voidp */);
+    if ((err == nullptr) && (txn_flags & TRITONSERVER_TXN_DECOUPLED) != 0) {
+      err = TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_UNSUPPORTED,
+          "ModelInfer RPC doesn't support models with decoupled "
+          "transaction policy");
+    }
+  }
+
+  // Create the inference request which contains all the
+  // input information needed for an inference.
+  TRITONSERVER_InferenceRequest* irequest = nullptr;
+  if (err == nullptr) {
+    err = TRITONSERVER_InferenceRequestNew(
+        &irequest, tritonserver_.get(), request.model_name().c_str(),
+        requested_model_version);
+  }
+
+  if (err == nullptr) {
+    err = SetInferenceRequestMetadata(irequest, request);
+  }
+
+  if (err == nullptr && header_forward_pattern_ != "") {
+    const auto& metadata = state->context_->ctx_->client_metadata();
+    for (const auto& pair : metadata) {
+      auto& key = pair.first;
+      auto& value = pair.second;
+      std::regex base_regex(header_forward_pattern_);
```

Comment:
Same as other comment, I think this line is expensive and can be improved by constructing the `std::regex` once and re-using that `std::regex` object

Reply:
Changed to prefix string matching 👍